hs/CalendarResponse.hs

```diff
@@ -0,0 +1,284 @@
+module CalendarResponse where
+
+import Data.List (sort, groupBy, sortBy)
+import Data.List.Split (splitOn)
+import Data.Time (Day, addDays, formatTime, getCurrentTime)
+import Happstack.Server (ServerPart, Response, toResponse)
+import Control.Monad.IO.Class (liftIO)
+import System.Locale
+import Database.CourseQueries (returnTutorialTimes, returnLectureTimes)
+import qualified Data.Text as T
+import Text.Read (readMaybe)
+import Database.Tables (Time, timeField)
+import Config (firstMondayFall,
+               lastWednesdayFall,
+               firstMondayWinter,
+               lastMondayWinter,
+               outDay,
+               fallHolidays,
+               winterHolidays)
+
+-- | Returns an ICS file of events as requested by the user.
+calendarResponse :: String -> ServerPart Response
+calendarResponse courses =
+    liftIO $ getCalendar courses
+
+-- | Gets together all the pieces of the program.
+getCalendar :: String -> IO Response
+getCalendar courses = do
+    let courseInfo = getCoursesInfo courses
+    databaseInfo <- mapM pullDatabase courseInfo
+    currentTime <- getCurrentTime
+    let systemTime = formatTime defaultTimeLocale "%Y%m%dT%H%M%SZ" currentTime
+    let events = databaseInfo >>= getEvent systemTime
+    return $ toResponse $ getICS events
+
+-- | Generates a string representing a ICS file.
+getICS :: [String] -> String
+getICS [] = ""
+getICS events = unlines $ header ++ events ++ bottom
+    where
+        header = ["BEGIN:VCALENDAR",
+                  "VERSION:2.0",
+                  "PRODID:-//Courseography//Calendar",
+                  "CALSCALE:GREGORIAN",
+                  "METHOD:PUBLISH"]
+        bottom = ["END:VCALENDAR"]
+
+-- ** Cookie Information
+
+-- | Obtains a list with all the information about the courses obtained from
+-- the cookies.
+getCoursesInfo :: String -> [(String, String, String)]
```

Comment:
Thank you @Ian-Stewart-Binks. I agree that creating type synonyms can help us gain a deeper understanding of the code. Moreover, I will be happy to work on writing new function signatures.
For this specific case getCoursesInfo :: String -> [(String, String, String)] could be replaced by getCoursesInfo :: String -> [(Code, Section, Session)].